tuber/server.py

```diff
@@ -40,11 +40,60 @@ def resolve_method(method):
     sig = None
 
     try:
-        sig = str(inspect.signature(method))
+        sig = inspect.signature(method)
     except:
-        pass
+        # pybind docstrings include a signature as the first line
+        if doc and doc.startswith(method.__name__ + "("):
+            if "\n" in doc:
+                sig, doc = doc.split("\n", 1)
+                doc = doc.strip()
+            else:
+                sig = doc
+                doc = None
+            sig = "(" + sig.split("(", 1)[1]
+
+    out = dict(__doc__=doc)
+
+    if isinstance(sig, str):
+        try:
+            # build a dummy function to parse its signature with inspect
+            code = compile(f"def sigfunc{sig}:\n pass", "sigfunc", "single")
+            exec(code, globals())
+            sig = inspect.signature(sigfunc)
+            params = list(sig.parameters.values())
+            p0 = params[0] if len(params) else None
+            # add self argument for unbound method
+            if not p0 or p0.name != "self":
+                if p0 and p0.kind == inspect.Parameter.POSITIONAL_ONLY:
+                    kind = p0.kind
+                else:
+                    kind = inspect.Parameter.POSITIONAL_OR_KEYWORD
+                parself = inspect.Parameter("self", kind)
+                sig = sig.replace(parameters=[parself] + params)
+        except:
+            sig = None
+
+    if sig is not None:
+        # parse signature parameters
+        annotations = {}
```

Comment:
I think this relies on dictionaries being ordered - which they are in Python, but not in JSON.

Reply:
Shouldn't matter for keyword arguments?  Could set the argument type to KEYWORD_ONLY for those to make sure there's no confusion.

Reply:
I think you're separating arguments into two categories:

- ordered arguments don't have defaults, and go into args
- named arguments have defaults and go into kwargs

In Python, this distinction is not part of the function definition - it's actually specified by the call site. The function prototype itself is fairly unopinionated, with two exceptions:

1. keyword-only arguments (`def foo(x, /, this_kind_of_thing): ...`)
2. *args and **kwargs as arguments

Unless I'm misunderstanding - I think any argument that gets dropped into "kwargs" is in principle unsafe (*) to use positionally (because JSON objects are unordered). At least in our codebase, we do this fairly casually and I don't think making it more restrictive is well justified.

(*) unsafe in a very limited way, I guess - we've constructed a client-side signature that may get the order wrong. Yup, this is contrived, but I think it's a structure that we can't rely on.

Reply:
Instead, why not create a list of arguments?

```
[ { "name": "arg1", "default": 3, "type": "float" }, ... ]
```

This is actually pretty close to what inspect.signature gives you anyways - there's no need to sort into args/kwargs buckets on the server side.

Reply:
Sure, although I'd store: name, default and optionally a stringified annotation (instead of type).